text/0000-macro-future-proofing.md

```diff
@@ -0,0 +1,141 @@
+};
+```
+
+Attempts to invoke this MBE will never succeed, because the macro parser
+will always emit an ambiguity error rather than make a choice when presented
+an ambiguity. In particular, it needs to decide when to stop accepting
+expressions for `foo` and look for a block for `bar` (noting that blocks are
+valid expressions). Situations like this are inherent to the macro system. On
+the other hand, it's possible to write an unambiguous matcher that becomes
+ambiguous due to changes in the syntax for the various fragments. As a
+concrete example:
+
+```rust
+macro_rules! bar {
+    ($in:ty ( $($arg:ident)*, ) -> $out:ty;) => (/*...*/)
+};
+```
+
+When the type syntax was extended to include the unboxed closure traits,
+an input such as `FnMut(i8, u8) -> i8;` became ambiguous. The goal of this
+proposal is to prevent such scenarios in the future by requiring certain
+"delimiter tokens" after an NT. When extending Rust's syntax in the future,
+ambiguity need only be considered when combined with these sets of delimiters,
+rather than any possible arbitrary matcher.
+
+# Detailed design
+
+The algorithm for recognizing valid matchers `M` follows. Note that a matcher
+is merely a token tree. A "simple NT" is an NT without repetitions. That is,
+`$foo:ty` is a simple NT but `$($foo:ty)+` is not. `FOLLOW(NT)` is the set of
+allowed tokens for the given NT's fragment specifier, and is defined below.
+`F` is used for representing the separator in complex NTs.  In `$($foo:ty),+`,
+`F` would be `,`, and for `$($foo:ty)+`, `F` would be `EOF`.
+
+*input*: a token tree `M` representing a matcher and a token `F`
+
+*output*: whether M is valid
+
+For each token `T` in `M`:
+
+1. If `T` is not an NT, continue.
+2. If `T` is a simple NT, look ahead to the next token `T'` in `M`. If
+   `T'` is `EOF` or a close delimiter of a token tree, replace `T'` with
+   `F`. If `T'` is in the set `FOLLOW(NT)`, `T'` is EOF, or `T'` is any close
+   delimiter, continue. Otherwise, reject.
+3. Else, `T` is a complex NT.
+    1. If `T` has the form `$(...)+` or `$(...)*`, run the algorithm on the
+       contents with `F` set to the token following `T`. If it accepts,
+       continue, else, reject.
+    2. If `T` has the form `$(...)U+` or `$(...)U*` for some token `U`, run
+       the algorithm on the contents with `F` set to `U`. If it accepts,
+       check that the last token in the sequence can be followed by `F`. If
+       so, accept. Otherwise, reject.
+
+This algorithm should be run on every matcher in every `macro_rules`
+invocation, with `F` as `EOF`. If it rejects a matcher, an error should be
+emitted and compilation should not complete.
+
+The current legal fragment specifiers are: `item`, `block`, `stmt`, `pat`,
+`expr`, `ty`, `ident`, `path`, `meta`, and `tt`.
+
+- `FOLLOW(pat)` = `{FatArrow, Comma, Eq}`
+- `FOLLOW(expr)` = `{FatArrow, Comma, Semicolon}`
+- `FOLLOW(ty)` = `{Comma, FatArrow, Colon, Eq, Gt, Ident(as)}`
```

Comment:
No, that was indeed an oversight

sent from my phone
On Jan 23, 2015 12:24 AM, "Vadim Chugunov" <notifications@github.com> wrote:

> In text/0000-macro-future-proofing.md
> <https://github.com/rust-lang/rfcs/pull/550#discussion_r23432570>:
>
> > +       continue, else, reject.
> > +    2. If `T` has the form `$(...)U+` or `$(...)U*` for some token `U`, run
> > +       the algorithm on the contents with `F` set to `U`. If it accepts,
> > +       check that the last token in the sequence can be followed by `F`. If
> > +       so, accept. Otherwise, reject.
> > +
> > +This algorithm should be run on every matcher in every `macro_rules`
> > +invocation, with `F` as `EOF`. If it rejects a matcher, an error should be
> > +emitted and compilation should not complete.
> > +
> > +The current legal fragment specifiers are: `item`, `block`, `stmt`, `pat`,
> > +`expr`, `ty`, `ident`, `path`, `meta`, and `tt`.
> > +
> > +- `FOLLOW(pat)` = `{FatArrow, Comma, Eq}`
> > +- `FOLLOW(expr)` = `{FatArrow, Comma, Semicolon}`
> > +- `FOLLOW(ty)` = `{Comma, FatArrow, Colon, Eq, Gt, Ident(as)}`
>
> @cmr <https://github.com/cmr>, was there a reason for not including
> Semicolon on this list as well?
> It's handy when defining matchers like $( fn $n:ident( $(p:ident :
> $t:ty),* ) -> $r:ty ; )* (note the trailing 'ty')
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/550/files#r23432570>.
>